src/source/Listener/GUDButtonListener.cpp

```diff
@@ -36,28 +40,31 @@ void GUDButtonListener::onButton1MouseClick(BaseButtonInstance* button)
 		}
 		else if(button->name == "Group")
 		{
-			GroupInstance * inst = new GroupInstance();
-			inst->setParent(g_dataModel->getWorkspace());
-			for(size_t i = 0; i < g_dataModel->getSelectionService()->getSelection().size(); i++)
-			{
-				if(g_dataModel->getSelectionService()->getSelection()[i]->canDelete)
+			if (g_dataModel->getSelectionService()->getSelection().size() > 1){
+				GroupInstance * inst = new GroupInstance();
+				inst->setParent(g_dataModel->getWorkspace());
+				for(size_t i = 0; i < g_dataModel->getSelectionService()->getSelection().size(); i++)
 				{
-					g_dataModel->getSelectionService()->getSelection()[i]->setParent(inst);
-					if(PartInstance* part = dynamic_cast<PartInstance*>(g_dataModel->getSelectionService()->getSelection()[i]))
+					if(g_dataModel->getSelectionService()->getSelection()[i]->canDelete)
 					{
-						inst->primaryPart = part;
+						g_dataModel->getSelectionService()->getSelection()[i]->setParent(inst);
+						if(PartInstance* part = dynamic_cast<PartInstance*>(g_dataModel->getSelectionService()->getSelection()[i]))
+						{
+							inst->primaryPart = part;
+						}
 					}
 				}
-			}
-			g_dataModel->getSelectionService()->clearSelection();
-			g_dataModel->getSelectionService()->addSelected(inst);
+				g_dataModel->getSelectionService()->clearSelection();
+				g_dataModel->getSelectionService()->addSelected(inst);
+			}			
 		}
 		else if(button->name == "UnGroup")
 		{
 			std::vector<Instance*> newinst;
 			for(size_t i = 0; i < g_dataModel->getSelectionService()->getSelection().size(); i++)
 			{
-				if(g_dataModel->getSelectionService()->getSelection()[i]->canDelete)
+				Instance* selection = g_dataModel->getSelectionService()->getSelection()[i];
+				if(selection->canDelete && selection->getClassName() == "GroupInstance")
```

Comment:
This check is useless as dynamic casting is done below, and wrong as comparing class names doesn't care about parent types, line 69 already does this

Reply:
You are correct, I have made the proper edits to fix this, would you also mind reviewing my model insertion code that I also implemented? It's a prototype but I'm not sure if my implementation is heading in the right direction or not.